### PR TITLE
[vm] Enable password and ssh key authentication during linux VM creation

### LINKED ISSUE #6230: az vm create doesn't work if the user specifies an admin password + ssh key, AND gives a useless error message
`$ az vm create --name=colemick-nixosdev1 --resource-group=colemick-nixosdev1 --admin-username=cole --admin-password=redacted '--ssh-key-value=...' --size=Standard_D16s_v3 --image=/subscriptions/aff271ee-e9be-4441-b9bb-42f5af4cbaeb/resourceGroups/aznixos/providers/Microsoft.Compute/images/nixos-image-16.09.1694.019dcc3-x86_64-linux-vhd`

`az vm create: error: incorrect usage for authentication-type 'password': [--admin-username USERNAME] --admin-password PASSWORD`

and then it works if I removed the SSH key.

1) The error message is completely unhelpful.
2) This should be allowed...

### PR DESCRIPTION
**Pull request by adewaleo**
Fixes #6230 

- `--authentication-type` now accepts all, which enables both password and ssh key authentication for linux vm's. Providing both password and ssh key value enables both forms of authentication without explicitly setting `--authentication-type`.
- Updated authentication type validation error message. 
- Added tests.

 
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py [hunk 1]
```diff
@@ -244,7 +244,7 @@ def build_msi_role_assignment(vm_vmss_name, vm_vmss_resource_id, role_definition
 def build_vm_resource(  # pylint: disable=too-many-locals
         cmd, name, location, tags, size, storage_profile, nics, admin_username,
         availability_set_id=None, admin_password=None, ssh_key_value=None, ssh_key_path=None,
-        image_reference=None, os_disk_name=None, custom_image_os_type=None,
+        image_reference=None, os_disk_name=None, custom_image_os_type=None, authentication_type=None,
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py]: @adewaleo, i suggest we don't create a new arg, rather only enable it when a user provides both password and ssh keys.

**Comment by adewaleo** [comment]
@yugangw-msft, please let me know if this looks good to you.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 2]
```diff
@@ -855,14 +854,36 @@ def _validate_vm_vmss_create_auth(namespace):
...
             namespace.ssh_dest_key_path = \
                 '/home/{}/.ssh/authorized_keys'.format(namespace.admin_username)
 
+    elif namespace.authentication_type == 'all':
+        if namespace.os_type.lower() == 'windows':
+            raise CLIError('SSH not supported for Windows VMs. Use password authentication.')
+
+        if not namespace.admin_password:
+            _prompt_for_password(namespace)
+        _validate_admin_password(namespace.admin_password,
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: NIT: make 2 parameters in the same line

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 3]
```diff
@@ -855,14 +854,36 @@ def _validate_vm_vmss_create_auth(namespace):
...
+                                 namespace.os_type)
+
+        validate_ssh_key(namespace)
+        if not namespace.ssh_dest_key_path:
+            namespace.ssh_dest_key_path = \
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: Unless it is 120 characters plus, make it same line.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_defaults.py [hunk 4]
```diff
@@ -299,9 +322,33 @@ def test_linux_with_password(self):
...
+        ns.authentication_type = 'all'
+        ns.admin_username = 'user12345'
+        ns.admin_password = 'verySecret!!!'
+        ns.ssh_key_value = self._get_test_ssh_key()
+
+        _validate_vm_vmss_create_auth(ns)
+        self.assertEqual(ns.authentication_type, 'all')
+        self.assertEqual(ns.ssh_dest_key_path, '/home/{}/.ssh/authorized_keys'.format(ns.admin_username))
+
+    def test_linux_with_password_and_ssh_explicit(self):
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_defaults.py]: this is a dupe test :)

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py (hunk 1 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py]: We decided to leave the arg in, after our discussion. Resolving this.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_defaults.py (hunk 4 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_defaults.py]: Hello Yugang, I will rename the test. The only difference is that the second test sets authentication_type to none

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py (hunk 3 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: I agree. I copied and pasted code from the other blocks without updating. Will make the changes.

**Pull request merged**